File: backtracking/1593_split_a_string_into_the_max_number_of_unique_substrings/1593_split_a_string_into_the_max_number_of_unique_substrings.go
package main

import "fmt"

func maxUniqueSplit(s string) int {
	return helper(s, map[string]bool{})
}

func helper(s string, seen map[string]bool) int {
	ans := 0
	for i := 1; i <= len(s); i++ {
		if !seen[s[0:i]] {
			seen[s[0:i]] = true
			ans = max(ans, 1+helper(s[i:], seen))
			seen[s[0:i]] = false
		}
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	s := "ababccc"
	res := maxUniqueSplit(s)
	fmt.Println(res)
}

File: backtracking/1593_split_a_string_into_the_max_number_of_unique_substrings/1593_split_a_string_into_the_max_number_of_unique_substrings.py

class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        self.ans = 0
        d = set()

        def helper(num, sub):
            if not sub:
                self.ans = max(num, self.ans)
                return
            
            if sub in d: return 
            
            for i in range(1, len(sub) + 1):
                if sub[:i] not in d:
                    d.add(sub[:i])
                    helper(num + 1, sub[i:])
                    d.remove(sub[:i]) 
                    
        helper(0, s)

        return self.ans



s = "ababccc"
res = Solution().maxUniqueSplit(s)
print(res)